### ISSUE TITLE
Improve Orbital Instability Guess

### ISSUE THREAD

**Issue by JonathonMisiewicz** [psi4/psi4#3045]
Right now, our SCF instability following algorithm is the following, very dumb one:
* Perform SCF
* Check instability
* If unstable, rotate in the direction of the unstable vector, according to some user-specified magnitude, then repeat
* If unstable _with a new eigenvalue_, increment the magnitude. It must have been bad.

We can guess more intelligently. A _smarter_ algorithm would look like:
* Perform SCF
* Check instability
* If unstable, rotate in the direction of the unstable vector, according to some user-specified magnitude, then...
* Do a line fit of the SCF energy in the direction of the unstable vector. The instability analysis gave us the first two terms of the Taylor expansion, analytically. Given that, we can then use the SCF energy at the rotated build to fit the _third_ derivative. It may also be possible to project the SCF gradient into the unstable direction. We can then use both pieces of information to fit the third and fourth derivative. (Some numerical tests will be necessary to determine whether the fourth dimensional fit is worthwhile, and good rotation magnitudes.)
* We then find the zeroes of f'(x)/x. This is a quadratic, and we can solve the roots analytically. If both are real, evaluate f at both zeroes to find a guess for the true minimum, and continue SCF at the guess rotation. If both are imaginary, a degree four polynomial fit predicts that f(x) has no zeroes, and our fit is unreliable. Print a warning and the current orbitals.

**Comment by JonathonMisiewicz** [comment]
I think you're confusing NR-root finding algorithms with instability following. Your answer makes sense only in the context of an different problem.

In the instability following problem, we start at a saddle point and compute both a negative eigenvalue of the orbital Hessian and the corresponding eigenvector. The eigenvector gives us the search direction. In the search direction, an orbital instability analysis alone gives us a a second-order Taylor expansion of the energy that is in the shape of a parabola. We are starting at its _maximum_, and it possesses no minimum. The eigenvalue gives us the parabola steepness, but neither I nor Psi4 have any idea how to use that information to give us a step size. Hence the need for a user-supplied step length at present.

The proposal here is that we use the energy and gradient information at another point along the line to fit the third and potentially fourth terms of the Taylor expansion, so we get a function that has a chance of having a minimum.

**Comment by TiborGY** [comment]
How about something like this:

- Perform SCF
- Check instability
- If unstable, rotate in the direction of the most unstable vector, according to some user-specified magnitude then...
- Compute the deviation of the new SCF energy from the 2nd order Taylor prediction, as well as the new SCF gradient
- If the 2nd order Taylor prediction was accurate enough (within some user-specified tolerance), and the magnitude of the gradient is within a comfortable range (another pair of parameters sadly...), then start SOSCF which should hopefully start going downhill.
- If the gradient is too low, increase the rotation magnitude, if too large decrease it. Recompute the new SCF energy and gradient, repeat the checks above.
- If the 2nd order Taylor prediction was inaccurate, decrease rotation magnitude. Recompute the new SCF energy and gradient, repeat the checks above.
- If the 2nd order Taylor prediction was inaccurate, **and** the gradient is too low, accept the orbital rotation, recompute the orbital Hessian at the new location (which is no longer exactly a saddle point), and repeat the entire follow process, until we find a point we can start SOSCF from.

This could of course get quite expensive if a lot of Hessian recomputes are required, and relies on SOSCF not climbing back up to the saddle point, but to me it sounds easier to implement.

With that said, I cannot help but feel like we are trying to reinvent a wheel. Surely, one would think that other people have already spent a lot of time thinking about the best ways to climb down from saddle points.

**Comment by TiborGY** [comment]
1. I was thinking about the broader optimization literature, not just SCF instabilities. The closest analogy to this problem may be where one is trying to start an IRC path from a saddle point, so the geometry optimization literature might have some ideas.
The machine learning folks have also been thinking about saddle point avoidance. [This paper](https://proceedings.neurips.cc/paper_files/paper/2014/file/17e23e50bedc63b4095e3d8204ce063b-Paper.pdf) discusses a generalized trust region method (they call it the "saddle-free Newton"), which may be interesting, although it is more about not converging to saddle points, rather than starting out from them.

2. I was thinking about how trust region methods adjust their step size based on the mismatch between predicted and actual change in function value. I think in sufficiently flat regions numerical issues can hamstring any optimization method.

3. I think that is subjective to some extent, for me it sounded conceptually simpler. Amount of new code required is probably comparable.

**Comment by JonathonMisiewicz** [comment]
I honestly don't recognize what you're trying to contribute. Are you proposing a course of action? Are you listing potentially relevant considerations for a more robust solution? Are you trying to say that "algorithms that are both more robust and harder to implement exist"?

**Comment by TiborGY** [comment]
> I honestly don't recognize what you're trying to contribute. Are you proposing a course of action? Are you listing potentially relevant considerations for a more robust solution? Are you trying to say that "algorithms that are both more robust and harder to implement exist"?

I was trying to bounce some ideas off the community, in the hope that it would ultimately lead to a better solution. To some extent that included all three of the above. I am sorry if that was frustrating, will try to clarify.

In my first post:
1. I proposed an alternative algorithm, which I guess is a course of action.
2. I highlighted two potential problems with my alternative algorithm.
3. I suggested that looking for existing algorithms in the literature may lead to something better than any of the proposals here, mine included.

In my second post:
1. I interpreted your response as "SCF instability literature is unlikely to have anything better". So I tried to clarify that I meant that the broader literature might contain useful ideas, not just SCF instability literature, but also the geometry optimization literature and possibly even the machine learning literature.
2. I suggested a particular paper and algorithm, which claims to be able to minimize functions effectively, while avoiding saddle points. This is the "both more robust and harder to implement" part. The idea that I should have explicitly written out, is that if it could be implemented as an SOSCF method, it may greatly reduce the chances of getting an unstable solution in the first place, which would make having smart SCF instability following less important.
3. I was trying to explain what thought process has lead me to believe that using the predicted/actual SCF energy may be a good idea for checking the quality of the Taylor expansion.

I hope no offense was caused.

**Comment by susilehtola** [comment]
Right, I guess I was a bit tired. The Taylor expansion of the energy in the orbital rotation

$$ E(\boldsymbol{\theta})=E_{0}+\boldsymbol{\theta}\cdot\boldsymbol{g}+\frac{1}{2}\boldsymbol{\theta}\cdot\boldsymbol{H}\cdot\boldsymbol{\theta} + \mathcal{O}(\boldsymbol{\theta}^3) $$

where 

$$ \boldsymbol{g} = \nabla_{\boldsymbol{\theta}}E\Big|_{\boldsymbol{\theta}=\boldsymbol{0}} $$

is the orbital gradient and

$$ \boldsymbol{H} = \nabla_{\boldsymbol{\theta}}\nabla_{\boldsymbol{\theta}}E\Big|_{\boldsymbol{\theta}=\boldsymbol{0}} $$

is the orbital Hessian reduces to 

$$ E(\boldsymbol{\theta}) \approx E_{0} + \frac{1}{2}\boldsymbol{\theta}\cdot\boldsymbol{H}\cdot\boldsymbol{\theta} $$

around a SCF solution determined by $\boldsymbol{g}=\boldsymbol{0}$. Diagonalizing $\boldsymbol{H}$ only gives you directions $\boldsymbol{\theta}$ in which the energy goes up or down, and you need to conduct a line search to find the minimum along a direction of descent.

The line search is pretty simple to carry out, though; you can use the quasiperiodicity of the orthogonal rotation matrix to determine a good trial step size, see e.g. our work on self-interaction corrected calculations in [J. Chem. Theory Comput. 12, 3195 (2016)](https://doi.org/10.1021/acs.jctc.6b00347). In my experience, a trial step of 1/5 of the shortest period of oscillation (see [Abrudan et al](https://doi.org/10.1016/j.sigpro.2009.03.015)) works well; this is slightly shorter than the theoretical value of 1/4: since you know that the function value is going down at zero step length (it is a descent direction), the minimum would be at pi/2 which is 1/4. Because of non-linear effects, the actual step should be shorter.

Employing this trick is key to achieving robust convergence for orbital rotation techniques, and will be used throughout in my upcoming orbital optimizer library.

**Comment by JonathonMisiewicz** [comment]
@TiborGY: I took no offense, just confusion. Thanks for clarifying! Improved SCF algorithms that _avoid saddle points in the first place_ would be great, but is outside the scope of this issue.
@susilehtola: Thanks. I can look into those papers before implementing anything. It may be a couple weeks, as I have some other priorities right now.
